File: app/src/main/java/com/shuneault/netrunnerdeckbuilder/db/DeckRepository.java
package com.shuneault.netrunnerdeckbuilder.db;

import com.shuneault.netrunnerdeckbuilder.game.Card;
import com.shuneault.netrunnerdeckbuilder.game.Deck;
import com.shuneault.netrunnerdeckbuilder.game.Format;

import java.util.ArrayList;

public class DeckRepository implements IDeckRepository {

    private final DatabaseHelper dbHelper;
    private final CardRepository cardRepo;
    private final ArrayList<Deck> mDecks;

    public DeckRepository(CardRepository cardRepo, DatabaseHelper dbHelper){
        this.cardRepo = cardRepo;
        this.dbHelper = dbHelper;
        mDecks = dbHelper.getAllDecks(true, this.cardRepo);
    }

    @Override
    public Deck getDeck(long deckId) {
        for (Deck deck : this.mDecks) {
            if (deck.getRowId() == deckId) {
                return deck;
            }
        }
        return null;
    }

    @Override
    public ArrayList<Deck> getAllDecks() {
        return mDecks;
    }

    @Override
    public void addDeck(Deck deck) {
        mDecks.add(deck);
    }

    @Override
    public boolean deleteDeck(Deck deck) {
        dbHelper.deleteDeck(deck);
        return mDecks.remove(deck);
    }

    @Override
    public void changeIdentity(Deck deck, String identityCode) {
        Card identity = cardRepo.getCard(identityCode);
        Deck d = getDeck(deck.getRowId());
        d.setIdentity(identity);
        dbHelper.updateDeck(d);
    }

    @Override
    public Long cloneDeck(Deck deck) {
        Deck newDeck = Deck.fromJSON(deck.toJSON(), cardRepo);
        newDeck.setName(String.format("Copy of %1$s", newDeck.getName()));
        // Do not clone the cards to add and cards to remove
        newDeck.setCardsToAdd(new ArrayList<>());
        newDeck.setCardsToRemove(new ArrayList<>());

        // Save in the database
        doCreateDeck(newDeck);
        return newDeck.getRowId();
    }

    private void doCreateDeck(Deck newDeck) {
        dbHelper.saveDeck(newDeck);
        addDeck(newDeck);
    }

    @Override
    public void createDeck(Deck deck) {
        doCreateDeck(deck);
    }

    @Override
    public void setDeckFormat(Deck deck, Format format) {
        deck.setFormat(format);
        dbHelper.updateDeck(deck);
    }

    @Override
    public void saveDeck(Deck deck) {
        dbHelper.saveDeck(deck);
    }

    @Override
    public void updateDeck(Deck deck) {
        dbHelper.updateDeck(deck);
    }
}

File: app/src/main/java/com/shuneault/netrunnerdeckbuilder/db/DeckRepository.kt
package com.shuneault.netrunnerdeckbuilder.db

import androidx.lifecycle.MediatorLiveData
import androidx.lifecycle.MutableLiveData
import com.shuneault.netrunnerdeckbuilder.game.Deck
import com.shuneault.netrunnerdeckbuilder.game.Format
import com.shuneault.netrunnerdeckbuilder.util.DeckSort
import kotlin.collections.ArrayList

class DeckRepository(private val cardRepo: CardRepository, private val dbHelper: DatabaseHelper) : IDeckRepository {
    private val mLiveDeckData: MutableLiveData<ArrayList<Deck>> = MutableLiveData()

    private val mDecks: ArrayList<Deck> = dbHelper.getAllDecks(true, cardRepo)
    private val currentOrder = DeckSort.DATEASCENDING
    val decks = MediatorLiveData<List<Deck>>()

    init {
        mLiveDeckData.value = mDecks

        decks.addSource(mLiveDeckData){ result: List<Deck>? ->
            result?.let { decks.value = sortDecks(it, currentOrder)}
        }
    }

    private fun sortDecks(list: List<Deck>, currentOrder: Any): List<Deck>? {
        return list.sortedBy { it.name }
    }

    override fun getDeck(deckId: Long): Deck? {
        for (deck in mDecks) {
            if (deck.rowId == deckId) {
                return deck
            }
        }
        return null
    }

    override fun allDecksLiveData(): MutableLiveData<java.util.ArrayList<Deck>> {
        return mLiveDeckData
    }

    // this is used to export data, not display.
    override val allDecks: ArrayList<Deck>
        get() = mDecks

    override fun addDeck(deck: Deck) {
        mDecks.add(deck)
        refreshLiveData()
    }

    private fun refreshLiveData() {
        mLiveDeckData.value = mDecks
    }

    override fun deleteDeck(deck: Deck): Boolean {
        dbHelper.deleteDeck(deck)
        val result = mDecks.remove(deck)
        refreshLiveData()
        return result
    }

    override fun changeIdentity(deck: Deck, identityCode: String) {
        val identity = cardRepo.getCard(identityCode)
        deck.rowId?.let {
            val x = getDeck(it)
            x!!.identity = identity
            dbHelper.updateDeck(x)
        }
    }

    override fun cloneDeck(deck: Deck): Long {
        val newDeck = Deck.fromJSON(deck.toJSON(), cardRepo)
        newDeck.name = String.format("Copy of %1\$s", newDeck.name)
        // Do not clone the cards to add and cards to remove
        newDeck.cardsToAdd = ArrayList()
        newDeck.cardsToRemove = ArrayList()

        // Save in the database
        doCreateDeck(newDeck)
        return newDeck.rowId!!
    }

    private fun doCreateDeck(newDeck: Deck) {
        dbHelper.saveDeck(newDeck)
        addDeck(newDeck)
    }

    override fun createDeck(deck: Deck) {
        doCreateDeck(deck)
    }

    override fun setDeckFormat(deck: Deck, format: Format) {
        deck.format = format
        dbHelper.updateDeck(deck)
    }

    override fun saveDeck(deck: Deck) {
        dbHelper.saveDeck(deck)
        refreshLiveData()
    }

    override fun updateDeck(deck: Deck) {
        dbHelper.updateDeck(deck)
        refreshLiveData()
    }


}